yt/utilities/io_handler.py

```diff
@@ -12,6 +13,26 @@
 
 use_caching = 0
 
+FluidFieldType = str
+FluidFieldName = str
+FluidFieldTuple = typing.Tuple[FluidFieldType, FluidFieldName]
+
+ParticleFieldType = str
+ParticleFieldName = str
+ParticleFieldTuple = typing.Tuple[ParticleFieldType, ParticleFieldName]
```

Comment:
```suggestion
ParticleFieldTuple = Tuple[str, str]
```
or actually, since there is no semantic difference between `ParticleFieldTuple` and plain `FieldTuple`, I think we should keep only the latter

Reply:
So I kind of disagree, there *is* a semantic difference -- there's no *technical* difference since they both resolve the same, but the key item is the difference between the meaning.  The thought process that led to [`NewType`](https://docs.python.org/3/library/typing.html#typing.NewType) is relevant here, but we can accomplish this without runtime checks.

Reply:
I didn't know about NewType, but I note that it actually doesn't make any sort of check or transform at runtime
```python
from typing import NewType

MyStr = NewType("MyStr", str)

def foo(a:MyStr):
    b = MyStr(a)
    print(type(a), type(b))
    print(f"{b!r}")

foo("example")
foo(1)
foo(None)
```
prints
```
<class 'str'> <class 'str'>
'example'
<class 'int'> <class 'int'>
1
<class 'NoneType'> <class 'NoneType'>
None
```
It doesn't seem relevant to use here either, however, since it would make type checking much much harder as we'd need to use these new types explicitly everywhere.

Reply:
Anyway you're right that there's a semantic difference, I completely misused this word, sorry. Still, I am not convinced yet that introducing these type aliases for str is the right approach here and I'd rather do that later when we have more experience with typing and we can make an informed decision on what `ParticleFieldType` should be (a type alias ? a `Literal` type ? an actual subclass ?). I am trying to be particularly careful here since we'll be using typing completely in the dark as long as we don't turn on some kind of static type checker on PRs, but I don't think this can realistically happen overnight, so I'd rather keep type hints as simple as possible for now. I hope this sounds acceptable.

Reply:
> Anyway you're right that there's a semantic difference, I completely misused this word, sorry.

I apologize for the strength with which I corrected the usage.

> Still, I am not convinced yet that introducing these type aliases for str is the right approach here and I'd rather do that later when we have more experience with typing and we can make an informed decision on what `ParticleFieldType` should be (a type alias ? a `Literal` type ? an actual subclass ?). I am trying to be particularly careful here since we'll be using typing completely in the dark as long as we don't turn on some kind of static type checker on PRs, but I don't think this can realistically happen overnight, so I'd rather keep type hints as simple as possible for now. I hope this sounds acceptable.

Alright, I am coming around a tiny bit, but I still do want to make them as explicit as possible with as wordy of names for situations where it's unclear.  I do like the idea of communicating more information than just `str`.  But, it might not be the right thing to do, or it might be solving the problem in the wrong way.  So let's indeed make this part of a bigger discussion.

Do you think that we should allow this in, while marking an issue to revisit once the decision has been made, or revert to the simpler ones before it goes in?  My inclination is the former, not just because it aligns with my already-done-work, but because it would be easier to remove the information in the future than to re-add it.

Reply:
> I apologize for the strength with which I corrected the usage.

no apologies needed, I genuinely appreciate being corrected on human language :-)

> Do you think that we should allow this in, while marking an issue to revisit once the decision has been made, or revert to the simpler ones before it goes in? My inclination is the former, not just because it aligns with my already-done-work, but because it would be easier to remove the information in the future than to re-add it.

I'd be far less worried about this adding technical debt if the custom types were defined in a "private" module, and on that condition I would get behind your preferred approach !

Reply:
I've been working on outstanding suggestions on this PR. @neutrinoceros, in terms of possibly defining these custom types in a private module, now that `main` now has a `_typing` module, would you suggest we move these custom type definitions over there (after a merge)? 

In terms of how to actually resolve the custom type discussion: I don't feel strongly one way or the other... but one possibility would be to put in the simplifications suggested by @neutrinoceros but add some comments that add the context that @matthewturk and I had to work through to create these custom types. Meaning in this particular case:

```
ParticleFieldTuple = Tuple[str, str]  # (field type, field name)
```
or a less trivial case:

```
ParticleTypeFields = typing.Dict[str, typing.List[str]]  # maps a particle type to fields of that type
```

So that the typing remains minimal but we still explain the intent of these types?

Reply:
> now that main now has a _typing module, would you suggest we move these custom type definitions over there (after a merge)?

yes


> one possibility would be to put in the simplifications suggested by @neutrinoceros but add some comments that add the context that @matthewturk and I had to work through to create these custom types. 

that would work for me. I don't want you guys to throw away the information you gathered while working on this.
Keep in mind that, while defining `FluidFieldType = str` is harmless to static type-checking, it's arguably more obscure from the user standpoint.
See for instance how, in the following example, my IDE represents the signature of a function I'm trying to use

![Screenshot 2022-01-07 at 10 13 26](https://user-images.githubusercontent.com/14075922/148520885-884bb94e-ceec-4fa6-b683-1b2bd2faa17b.png)

One can see the alias type is not resolved (note that it's exactly the same situation with alternative approaches using `NewType`, `Literal`,  or `TypeAlias`). It can be argued that `FluidFieldType` better communicates the meaning of the variable, but at the same time it obscures the actual expected type, which I think is counter productive for a type annotation. Another drawback to this approach is that it doesn't encourage using clear variable names.

TBH I don't fully understand the use cases for `NewType` or `TypeAlias` (and I'm almost convinced `Literal` is inappropriate here), so I don't think we should commit to any complexity greater than minimal at first.

Reply:
> It can be argued that FluidFieldType better communicates the meaning of the variable, but at the same time it obscures the actual expected type, which I think is counter productive for a type annotation.

Yes! I think we were using the typing to add info that really should be in the comments (or if refactoring were easier we might rename some variables to be more descriptive). I'm now even more on the side of simplifying the typing we added while adding comments to better explain the function arguments... will give it a shot, hopefully will update soon.